### PR TITLE
feat: add support for contracts with events

### PR DESCRIPTION
**Pull request by xJonathanLEI**
Resolves task 4 of #27:

> reading contract artifacts with events

This PR:

- adds support for `starknet-compile` `v0.7.0` artifacts, which have an additional `attributes` field under `program`;
- adds a way to deterministically generate contract artifacts (with Docker) for test data;
- adds a new ABI entry type: `Event`.

### CODE REVIEW

**Review by milancermak — APPROVED** [review]
LGTM

🚢

File: starknet-contract/test-data/generate_artifacts.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# Deterministically generate contract artifacts
+
+docker run -it --rm \
+    -v "$(pwd)/artifacts:/artifacts" \
+    -v "$(pwd)/contracts:/contracts:ro" \
+    -v "$(pwd)/docker_entry.sh:/entry.sh:ro" \
+    --env "USER_ID=$(id -u)" \
+    --env "GROUP_ID=$(id -g)" \
+    --entrypoint "/entry.sh" \
+    shardlabs/cairo-cli:0.7.0
```
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply starknet-contract/test-data/generate_artifacts.sh:12]: I think so. That way we get to make sure the library can handle artifacts produced by the latest compiler version.

File: starknet-contract/src/artifact.rs [hunk 2]
```diff
@@ -17,6 +17,8 @@ pub struct Artifact {
 #[derive(Debug, Serialize, Deserialize)]
 #[serde(deny_unknown_fields)]
 pub struct Program {
+    #[serde(default, skip_serializing)]
+    pub attributes: serde::de::IgnoredAny, // Skipped since it's not used in deployment
```
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply starknet-contract/src/artifact.rs:21]: You're right. `default` is useless here. Removing.

File: starknet-core/src/types/contract_code.rs [hunk 3]
```diff
@@ -47,6 +48,13 @@ pub struct L1Handler {
     pub outputs: Vec<Output>,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Event {
+    pub name: String,
+    pub keys: Vec<()>, // Can't figure out what's in `keys`
```
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply starknet-core/src/types/contract_code.rs:54]: Yeah under the context of actually emitted events (i.e. `get_block` & `get_transaction_receipt`), `keys` contains exactly one element of `pedersen_hash(EVENT_NAME)`.

However, in the context of ABIs, there seems to be no way to generate ABI with non-empty `keys`. I tried compilation contracts with events, but it doesn't work.

File: starknet-contract/test-data/generate_artifacts.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (milancermak) [reply starknet-contract/test-data/generate_artifacts.sh:12]: Sounds good

**Pull request merged**